In [1]:
import sys

import pandas as pd



print("Python version:", sys.version)

print("Pandas version:", pd.__version__)

print("Kernel is working.")

Python version: 3.13.5 | packaged by Anaconda, Inc. | (main, Jun 12 2025, 16:37:03) [MSC v.1929 64 bit (AMD64)]
Pandas version: 2.2.3
Kernel is working.


In [2]:
from pathlib import Path

project_root = Path.cwd()

# If the notebook is running from the notebooks folder, move up to DataGuard
if project_root.name == "notebooks":
    project_root = project_root.parent

print("Current working folder:", Path.cwd())
print("Project root:", project_root)
print("Data folder exists:", (project_root / "data").exists())
print("Notebooks folder exists:", (project_root / "notebooks").exists())
print("Scripts folder exists:", (project_root / "scripts").exists())

Current working folder: c:\Users\samee\DataGuard\notebooks
Project root: c:\Users\samee\DataGuard
Data folder exists: True
Notebooks folder exists: True
Scripts folder exists: True


In [3]:
data_folder = project_root / "data"

print("Files inside data folder:")
for file in data_folder.iterdir():
    print(file.name)

Files inside data folder:
nyc_311_raw_100k.csv


In [4]:
data_path = project_root / "data" / "nyc_311_raw_100k.csv"

df = pd.read_csv(data_path)

print("Dataset loaded successfully.")
print("Rows and columns:", df.shape)

df.head()

Dataset loaded successfully.
Rows and columns: (100000, 44)


,unique_key,created_date,closed_date,agency,agency_name,complaint_type,descriptor,descriptor_2,location_type,incident_zip,...,vehicle_type,taxi_company_borough,taxi_pick_up_location,bridge_highway_name,bridge_highway_direction,road_ramp,bridge_highway_segment,latitude,longitude,location
0,69486831,2026-06-26T02:06:32.000,NaN,NYPD,New York City Police Department,Noise - Commercial,Loud Music/Party,NaN,Club/Bar/Restaurant,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,40.686278,-73.973707,POINT (-73.9737071634 40.686277751362)
1,69492768,2026-06-26T02:06:05.000,NaN,TLC,Taxi and Limousine Commission,For Hire Vehicle Complaint,Driver Complaint - Non Passenger,Unsafe Driving,Street,11211.0,...,NaN,NaN,"180 UNION AVENUE, BROOKLYN, NY, 11211",NaN,NaN,NaN,NaN,40.705761,-73.950259,POINT (-73.950259119641 40.705760603484)
2,69482343,2026-06-26T02:05:24.000,NaN,NYPD,New York City Police Department,Noise - Residential,Loud Music/Party,NaN,Residential Building/House,10013.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,40.717730,-73.998185,POINT (-73.998185472064 40.717730264942)
3,69488306,2026-06-26T02:03:49.000,NaN,NYPD,New York City Police Department,Noise - Street/Sidewalk,Loud Music/Party,NaN,Street/Sidewalk,11249.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,40.721489,-73.958332,POINT (-73.958332047132 40.721488549007)
4,69482692,2026-06-26T02:03:29.000,NaN,NYPD,New York City Police Department,Illegal Parking,Blocked Crosswalk,NaN,Street/Sidewalk,11421.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,40.689885,-73.858108,POINT (-73.858107579574 40.689884995274)


In [6]:
df.columns.tolist()

['unique_key',
 'created_date',
 'closed_date',
 'agency',
 'agency_name',
 'complaint_type',
 'descriptor',
 'descriptor_2',
 'location_type',
 'incident_zip',
 'incident_address',
 'street_name',
 'cross_street_1',
 'cross_street_2',
 'intersection_street_1',
 'intersection_street_2',
 'address_type',
 'city',
 'landmark',
 'facility_type',
 'status',
 'due_date',
 'resolution_description',
 'resolution_action_updated_date',
 'community_board',
 'council_district',
 'police_precinct',
 'bbl',
 'borough',
 'x_coordinate_state_plane',
 'y_coordinate_state_plane',
 'open_data_channel_type',
 'park_facility_name',
 'park_borough',
 'vehicle_type',
 'taxi_company_borough',
 'taxi_pick_up_location',
 'bridge_highway_name',
 'bridge_highway_direction',
 'road_ramp',
 'bridge_highway_segment',
 'latitude',
 'longitude',
 'location']

In [9]:
column_inventory = pd.DataFrame({
    "column_number": range(1, len(df.columns) + 1),
    "column_name": df.columns,
    "data_type": df.dtypes.values
})

column_inventory

,column_number,column_name,data_type
0,1,unique_key,int64
1,2,created_date,object
2,3,closed_date,object
3,4,agency,object
4,5,agency_name,object
5,6,complaint_type,object
6,7,descriptor,object
7,8,descriptor_2,object
8,9,location_type,object
9,10,incident_zip,float64


In [10]:
df.head()

,unique_key,created_date,closed_date,agency,agency_name,complaint_type,descriptor,descriptor_2,location_type,incident_zip,...,vehicle_type,taxi_company_borough,taxi_pick_up_location,bridge_highway_name,bridge_highway_direction,road_ramp,bridge_highway_segment,latitude,longitude,location
0,69486831,2026-06-26T02:06:32.000,NaN,NYPD,New York City Police Department,Noise - Commercial,Loud Music/Party,NaN,Club/Bar/Restaurant,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,40.686278,-73.973707,POINT (-73.9737071634 40.686277751362)
1,69492768,2026-06-26T02:06:05.000,NaN,TLC,Taxi and Limousine Commission,For Hire Vehicle Complaint,Driver Complaint - Non Passenger,Unsafe Driving,Street,11211.0,...,NaN,NaN,"180 UNION AVENUE, BROOKLYN, NY, 11211",NaN,NaN,NaN,NaN,40.705761,-73.950259,POINT (-73.950259119641 40.705760603484)
2,69482343,2026-06-26T02:05:24.000,NaN,NYPD,New York City Police Department,Noise - Residential,Loud Music/Party,NaN,Residential Building/House,10013.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,40.717730,-73.998185,POINT (-73.998185472064 40.717730264942)
3,69488306,2026-06-26T02:03:49.000,NaN,NYPD,New York City Police Department,Noise - Street/Sidewalk,Loud Music/Party,NaN,Street/Sidewalk,11249.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,40.721489,-73.958332,POINT (-73.958332047132 40.721488549007)
4,69482692,2026-06-26T02:03:29.000,NaN,NYPD,New York City Police Department,Illegal Parking,Blocked Crosswalk,NaN,Street/Sidewalk,11421.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,40.689885,-73.858108,POINT (-73.858107579574 40.689884995274)


In [11]:
df.tail()

,unique_key,created_date,closed_date,agency,agency_name,complaint_type,descriptor,descriptor_2,location_type,incident_zip,...,vehicle_type,taxi_company_borough,taxi_pick_up_location,bridge_highway_name,bridge_highway_direction,road_ramp,bridge_highway_segment,latitude,longitude,location
99995,69388945,2026-06-16T19:46:25.000,2026-06-18T20:13:06.000,HPD,Department of Housing Preservation and Develop...,HEAT/HOT WATER,APARTMENT ONLY,NO HOT WATER,RESIDENTIAL BUILDING,11692.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,40.597232,-73.803736,POINT (-73.803736314286 40.597231529209)
99996,69378101,2026-06-16T19:46:25.000,2026-06-17T01:29:02.000,NYPD,New York City Police Department,Noise - Residential,Loud Music/Party,NaN,Residential Building/House,11222.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,40.737045,-73.955396,POINT (-73.955395859868 40.737044726028)
99997,69387434,2026-06-16T19:46:25.000,2026-06-18T20:13:06.000,HPD,Department of Housing Preservation and Develop...,APPLIANCE,ELECTRIC/GAS RANGE,OVEN DOOR MISSING OR BROKEN,RESIDENTIAL BUILDING,11692.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,40.597232,-73.803736,POINT (-73.803736314286 40.597231529209)
99998,69379069,2026-06-16T19:46:22.000,2026-06-16T19:49:31.000,NYPD,New York City Police Department,Illegal Parking,License Plate Obscured,NaN,Street/Sidewalk,10033.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,40.850297,-73.935671,POINT (-73.93567094185 40.850296679687)
99999,69383241,2026-06-16T19:46:09.000,NaN,DOT,Department of Transportation,Curb Condition,Curb Defect-Metal Protruding,NaN,Sidewalk,11218.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,40.635673,-73.966640,POINT (-73.966640048487 40.635673163893)


In [14]:
df.dtypes

unique_key                          int64
created_date                       object
closed_date                        object
agency                             object
agency_name                        object
complaint_type                     object
descriptor                         object
descriptor_2                       object
location_type                      object
incident_zip                      float64
incident_address                   object
street_name                        object
cross_street_1                     object
cross_street_2                     object
intersection_street_1              object
intersection_street_2              object
address_type                       object
city                               object
landmark                           object
facility_type                      object
status                             object
due_date                           object
resolution_description             object
resolution_action_updated_date    

Some text-like fields are stored as object columns, including complaint type, agency, status and borough. Several date fields are also currently stored as object columns, meaning they will need to be converted to datetime before time-based analysis. Location coordinates are stored as float64, which is appropriate for latitude and longitude. The incident_zip column is stored as float64 because of missing values, but ZIP codes are identifiers and may need to be treated as text later.

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 44 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   unique_key                      100000 non-null  int64  
 1   created_date                    100000 non-null  object 
 2   closed_date                     74905 non-null   object 
 3   agency                          100000 non-null  object 
 4   agency_name                     100000 non-null  object 
 5   complaint_type                  100000 non-null  object 
 6   descriptor                      97603 non-null   object 
 7   descriptor_2                    33222 non-null   object 
 8   location_type                   86839 non-null   object 
 9   incident_zip                    99189 non-null   float64
 10  incident_address                96281 non-null   object 
 11  street_name                     96281 non-null   object 
 12  cross_street_1   

## Dataset Info Observations

The dataset contains 100,000 rows and 44 columns. Most key business fields, such as `unique_key`, `created_date`, `agency`, `agency_name`, and `complaint_type`, are complete in this sample.

Several columns contain missing values. For example, `closed_date` is missing for some records, which may indicate requests that were still open or not closed at the time the data was pulled. Other fields, such as `descriptor_2`, `facility_type`, and some location-related fields, are also incomplete.

The dataset contains a mix of object, integer and float columns. Many text-like and date-like fields are currently stored as object columns. These will need further review in later stages of the DataGuard project.

In [17]:
missing_values_preview = df.isnull().sum().sort_values(ascending=False).head(15)

missing_values_preview

taxi_company_borough        99922
facility_type               99775
bridge_highway_direction    99744
road_ramp                   99743
due_date                    99574
bridge_highway_segment      99544
bridge_highway_name         99538
taxi_pick_up_location       98753
vehicle_type                94573
descriptor_2                66778
landmark                    33400
intersection_street_2       25207
intersection_street_1       25202
closed_date                 25095
cross_street_2              21268
dtype: int64

## Missing Values Preview

The missing values preview shows that some columns are missing values for most records. Many of the most incomplete columns appear to be conditional fields, meaning they only apply to specific types of 311 requests.

For example, taxi-related fields, vehicle fields and bridge/highway fields may only be relevant for certain complaint categories. Therefore, missing values in these columns are not automatically errors.

However, some missing values may require further investigation later. For example, `closed_date` is missing for 25,095 records, which may indicate requests that were still open or had no closure timestamp at the time the data was pulled.

For Day 1, I am only observing missing values. A deeper missing-value analysis will be completed later as part of the DataGuard data quality checks.

# Day 1 Summary

In this notebook, I completed the initial data discovery stage of the DataGuard project using a 100,000-row sample of the NYC 311 service requests dataset.

## Work Completed

- Verified the Python and pandas environment using the Anaconda kernel
- Loaded the NYC 311 dataset from the local `data` folder
- Confirmed the dataset contains 100,000 rows and 44 columns
- Created a column inventory showing column names and data types
- Previewed the first and last rows using `head()` and `tail()`
- Reviewed data types using `dtypes`
- Reviewed dataset structure, non-null counts, and memory usage using `info()`
- Previewed missing values across columns

## Key Observations

The dataset contains real-world 311 service request records, including complaint types, agencies, statuses, dates, location fields, and resolution-related information.

Several important fields such as `unique_key`, `created_date`, `agency`, `agency_name`, and `complaint_type` are complete in this sample.

Some columns have many missing values. Many of these appear to be conditional fields, such as taxi-related or bridge/highway-related columns. These missing values are not automatically errors because those fields only apply to certain complaint types.

Several date columns are currently stored as object/text columns and will need to be converted to datetime later for time-based analysis.

## Next Step

The next step is Day 2: use pandas to answer basic business questions about the dataset, such as the most common complaint types, borough-level complaint volume, agency workload, and request statuses.

## Skill 2: Counting Category Values with `value_counts()`

`value_counts()` counts how often each unique value appears in a column.

This is useful for identifying high-volume categories, such as the most common complaint types, boroughs with the most requests, agencies with the most workload, and request status distribution.

In [37]:
# Most common complaint types
df["complaint_type"].value_counts().head(10)

complaint_type
Illegal Parking            17228
Noise - Residential        11007
Noise - Street/Sidewalk     7746
Blocked Driveway            4965
UNSANITARY CONDITION        3430
Street Condition            3189
Water System                2104
Noise - Vehicle             2070
Dirty Condition             2066
Abandoned Vehicle           1979
Name: count, dtype: int64

### Observation: Most Common Complaint Types

The most common complaint type in this 100,000-row NYC 311 sample is `Illegal Parking`, with 17,228 requests.

Noise-related complaints also appear frequently, including `Noise - Residential`, `Noise - Street/Sidewalk`, and `Noise - Vehicle`.

This suggests that parking and noise issues are major sources of 311 service request volume in this sample.

In [38]:
# Complaint count by borough
df["borough"].value_counts()

borough
BROOKLYN         31657
QUEENS           25993
BRONX            19431
MANHATTAN        18810
STATEN ISLAND     3983
Unspecified        126
Name: count, dtype: int64

### Observation: Complaint Count by Borough

Brooklyn has the highest number of 311 requests in this sample, with 31,657 complaints. Queens has the second-highest volume, followed by the Bronx and Manhattan.

The dataset also contains 126 records with `Unspecified` borough values. These records may need further review during the data quality stage because borough is an important location field for analysis.

In [39]:
# Agencies with the most 311 requests
df["agency"].value_counts().head(10)

agency
NYPD     50751
HPD      13700
DSNY      8220
DOT       7322
DEP       5582
DPR       5368
DOB       3033
DOHMH     2301
DHS       1577
TLC       1277
Name: count, dtype: int64

### Observation: Agency Workload

The NYPD handles the highest number of 311 requests in this sample, with 50,751 records. This is much higher than the next agencies, including HPD, DSNY, DOT, and DEP.

This suggests that a large portion of the sampled 311 workload is related to issues routed to NYPD, such as illegal parking and noise complaints.

In [40]:
# Request status distribution
df["status"].value_counts()

status
Closed         72749
Open           14206
In Progress    12183
Started          448
Assigned         348
Pending           66
Name: count, dtype: int64

### Observation: Request Status Distribution

Most 311 requests in this sample are marked as `Closed`, with 72,749 records. However, there are also 14,206 `Open` requests and 12,183 `In Progress` requests.

This status distribution is useful because request status affects how the dataset should be interpreted. For example, missing `closed_date` values may be expected for requests that are still open or in progress.

## Skill 3: Filtering Rows

Filtering rows means keeping only the records that match a specific condition.

For the NYC 311 dataset, filtering is useful when analyzing subsets such as complaints from one borough, complaints handled by one agency, noise-related complaints, or requests with a specific status.

In [43]:
analysis_columns = [
    "created_date",
    "closed_date",
    "agency",
    "agency_name",
    "complaint_type",
    "descriptor",
    "borough",
    "status",
    "city"
]

print("analysis_columns created.")

analysis_columns created.


In [44]:
# Filter for Brooklyn complaints
brooklyn_df = df[df["borough"] == "BROOKLYN"]

print("Brooklyn complaints:", brooklyn_df.shape[0])

brooklyn_df[analysis_columns].head()

Brooklyn complaints: 31657


,created_date,closed_date,agency,agency_name,complaint_type,descriptor,borough,status,city
1,2026-06-26T02:06:05.000,NaN,TLC,Taxi and Limousine Commission,For Hire Vehicle Complaint,Driver Complaint - Non Passenger,BROOKLYN,In Progress,BROOKLYN
3,2026-06-26T02:03:49.000,NaN,NYPD,New York City Police Department,Noise - Street/Sidewalk,Loud Music/Party,BROOKLYN,In Progress,BROOKLYN
5,2026-06-26T02:02:57.000,NaN,NYPD,New York City Police Department,Illegal Parking,Blocked Bike Lane,BROOKLYN,In Progress,BROOKLYN
7,2026-06-26T02:01:10.000,NaN,NYPD,New York City Police Department,Noise - Residential,Loud Talking,BROOKLYN,In Progress,BROOKLYN
8,2026-06-26T02:00:52.000,NaN,NYPD,New York City Police Department,Illegal Parking,Blocked Bike Lane,BROOKLYN,In Progress,BROOKLYN


### Observation: Brooklyn Complaints

Filtering for `BROOKLYN` creates a smaller DataFrame containing only Brooklyn service requests. This makes it possible to analyze complaint patterns within one borough instead of the full citywide dataset.

In [45]:
# Filter for noise-related complaints
noise_df = df[df["complaint_type"].str.contains("Noise", case=False, na=False)]

print("Noise-related complaints:", noise_df.shape[0])

noise_df[analysis_columns].head()

Noise-related complaints: 24737


,created_date,closed_date,agency,agency_name,complaint_type,descriptor,borough,status,city
0,2026-06-26T02:06:32.000,NaN,NYPD,New York City Police Department,Noise - Commercial,Loud Music/Party,Unspecified,In Progress,NaN
2,2026-06-26T02:05:24.000,NaN,NYPD,New York City Police Department,Noise - Residential,Loud Music/Party,MANHATTAN,In Progress,NEW YORK
3,2026-06-26T02:03:49.000,NaN,NYPD,New York City Police Department,Noise - Street/Sidewalk,Loud Music/Party,BROOKLYN,In Progress,BROOKLYN
7,2026-06-26T02:01:10.000,NaN,NYPD,New York City Police Department,Noise - Residential,Loud Talking,BROOKLYN,In Progress,BROOKLYN
10,2026-06-26T02:00:20.000,NaN,NYPD,New York City Police Department,Noise - Residential,Banging/Pounding,BRONX,In Progress,BRONX


### Observation: Noise-Related Complaints

There are 24,737 noise-related complaints in this 100,000-row NYC 311 sample.

The filter searched for complaint types containing the word `Noise`, which includes categories such as `Noise - Commercial`, `Noise - Residential`, and `Noise - Street/Sidewalk`.

This subset will be useful for analyzing noise complaint patterns by borough, status and agency.

In [46]:
# Filter for noise-related complaints in Brooklyn
brooklyn_noise_df = df[
    (df["borough"] == "BROOKLYN") &
    (df["complaint_type"].str.contains("Noise", case=False, na=False))
]

print("Brooklyn noise-related complaints:", brooklyn_noise_df.shape[0])

brooklyn_noise_df[analysis_columns].head()

Brooklyn noise-related complaints: 7005


,created_date,closed_date,agency,agency_name,complaint_type,descriptor,borough,status,city
3,2026-06-26T02:03:49.000,NaN,NYPD,New York City Police Department,Noise - Street/Sidewalk,Loud Music/Party,BROOKLYN,In Progress,BROOKLYN
7,2026-06-26T02:01:10.000,NaN,NYPD,New York City Police Department,Noise - Residential,Loud Talking,BROOKLYN,In Progress,BROOKLYN
14,2026-06-26T01:59:05.000,NaN,NYPD,New York City Police Department,Noise - Residential,Loud Talking,BROOKLYN,In Progress,BROOKLYN
26,2026-06-26T01:55:40.000,NaN,NYPD,New York City Police Department,Noise - Commercial,Loud Music/Party,BROOKLYN,In Progress,BROOKLYN
28,2026-06-26T01:54:55.000,NaN,NYPD,New York City Police Department,Noise - Vehicle,Car/Truck Horn,BROOKLYN,In Progress,NaN


### Observation: Brooklyn Noise Complaints

Filtering with two conditions created a subset of records that are both located in `BROOKLYN` and noise-related.

This type of filtering is useful because many business questions are not citywide. Analysts often need to answer questions for a specific location, complaint category, agency or status.

## Skill 4: Sorting Values

Sorting helps arrange records or summary results in a meaningful order.

For NYC 311 analysis, sorting can help identify the newest requests, oldest requests, highest-volume complaint types, or highest-volume boroughs.

In [48]:
# Sort records by created_date from newest to oldest
df.sort_values(by="created_date", ascending=False)[analysis_columns].head()

,created_date,closed_date,agency,agency_name,complaint_type,descriptor,borough,status,city
0,2026-06-26T02:06:32.000,NaN,NYPD,New York City Police Department,Noise - Commercial,Loud Music/Party,Unspecified,In Progress,NaN
1,2026-06-26T02:06:05.000,NaN,TLC,Taxi and Limousine Commission,For Hire Vehicle Complaint,Driver Complaint - Non Passenger,BROOKLYN,In Progress,BROOKLYN
2,2026-06-26T02:05:24.000,NaN,NYPD,New York City Police Department,Noise - Residential,Loud Music/Party,MANHATTAN,In Progress,NEW YORK
3,2026-06-26T02:03:49.000,NaN,NYPD,New York City Police Department,Noise - Street/Sidewalk,Loud Music/Party,BROOKLYN,In Progress,BROOKLYN
4,2026-06-26T02:03:29.000,NaN,NYPD,New York City Police Department,Illegal Parking,Blocked Crosswalk,QUEENS,In Progress,NaN


## Skill 5: Converting Date Columns

The `created_date` and `closed_date` columns were loaded as object/text columns.

To perform reliable time-based analysis, these columns should be converted to datetime format. This allows pandas to correctly identify earliest dates, latest dates, and later calculate resolution time.

In [49]:
# Convert date columns from text/object to datetime
df["created_date"] = pd.to_datetime(df["created_date"], errors="coerce")
df["closed_date"] = pd.to_datetime(df["closed_date"], errors="coerce")

print("Date conversion complete.")
print(df[["created_date", "closed_date"]].dtypes)

Date conversion complete.
created_date    datetime64[ns]
closed_date     datetime64[ns]
dtype: object


In [50]:
# Convert date columns from text/object to datetime
df["created_date"] = pd.to_datetime(df["created_date"], errors="coerce")
df["closed_date"] = pd.to_datetime(df["closed_date"], errors="coerce")

print("Date conversion complete.")
print(df[["created_date", "closed_date"]].dtypes)

Date conversion complete.
created_date    datetime64[ns]
closed_date     datetime64[ns]
dtype: object


In [51]:
# Date range covered by the dataset
earliest_created_date = df["created_date"].min()
latest_created_date = df["created_date"].max()

print("Earliest created date:", earliest_created_date)
print("Latest created date:", latest_created_date)

Earliest created date: 2026-06-16 19:46:09
Latest created date: 2026-06-26 02:06:32


### Observation: Date Range

The 100,000-row sample covers requests created between June 16, 2026 and June 26, 2026.

This means the notebook is analyzing a recent sample of NYC 311 requests rather than the full historical dataset.

# Business Questions

This section answers 10 basic business questions about the NYC 311 dataset using pandas.

Each question includes a short business purpose, pandas code and a plain-English observation.

## Question 1: How many service requests are in this sample?

This question gives a basic sense of the dataset size. Each row represents one NYC 311 service request.

In [53]:
# Number of rows and columns in the dataset
num_rows = df.shape[0]
num_columns = df.shape[1]

print("Number of service requests:", num_rows)
print("Number of columns:", num_columns)

Number of service requests: 100000
Number of columns: 44


## Question 2: What are the most common complaint types?

This question identifies the highest-volume complaint categories in the NYC 311 sample. High-volume complaint types can show which city issues generate the most service requests.

In [54]:
# Top 10 most common complaint types
top_complaint_types = df["complaint_type"].value_counts().head(10)

top_complaint_types

complaint_type
Illegal Parking            17228
Noise - Residential        11007
Noise - Street/Sidewalk     7746
Blocked Driveway            4965
UNSANITARY CONDITION        3430
Street Condition            3189
Water System                2104
Noise - Vehicle             2070
Dirty Condition             2066
Abandoned Vehicle           1979
Name: count, dtype: int64

### Observation

The most common complaint type in this sample is `Illegal Parking`, followed by several noise-related categories such as `Noise - Residential`, `Noise - Street/Sidewalk`, and `Noise - Vehicle`.

This suggests that parking and noise issues are major sources of NYC 311 service request volume in this sample.

## Question 3: Which boroughs have the most 311 requests?

This question identifies which boroughs have the highest number of service requests in the sample. Borough-level analysis helps show where complaint volume is concentrated geographically.

In [55]:
# Complaint count by borough
borough_counts = df["borough"].value_counts()

borough_counts

borough
BROOKLYN         31657
QUEENS           25993
BRONX            19431
MANHATTAN        18810
STATEN ISLAND     3983
Unspecified        126
Name: count, dtype: int64

### Observation

Brooklyn has the highest number of 311 requests in this sample, with 31,657 complaints. Queens has the second-highest volume, followed by the Bronx and Manhattan.

There are also 126 records with `Unspecified` borough values. These records may need review later because borough is an important location field for analysis.

## Question 4: Which agencies handle the most 311 requests?

This question shows which NYC agencies are responsible for the highest number of service requests in the sample. This can help identify agency workload patterns.

In [56]:
# Top 10 agencies by number of requests
top_agencies = df["agency"].value_counts().head(10)

top_agencies

agency
NYPD     50751
HPD      13700
DSNY      8220
DOT       7322
DEP       5582
DPR       5368
DOB       3033
DOHMH     2301
DHS       1577
TLC       1277
Name: count, dtype: int64

### Observation

The NYPD handles the highest number of 311 requests in this sample, with 50,751 records. This is much higher than the next agencies, including HPD, DSNY, DOT, and DEP.

This suggests that a large share of the sampled 311 workload is related to complaint types routed to NYPD, such as illegal parking and noise complaints.

## Question 5: What are the most common request statuses?

This question shows the distribution of request statuses in the sample. Status analysis helps understand how many requests are closed, open, in progress, or pending.

In [58]:
# Request status distribution
status_counts = df["status"].value_counts()

status_counts

status
Closed         72749
Open           14206
In Progress    12183
Started          448
Assigned         348
Pending           66
Name: count, dtype: int64

### Observation

Most requests in this sample are marked as `Closed`, with 72,749 records. However, there are also 14,206 `Open` requests and 12,183 `In Progress` requests.

This is important because request status affects how we interpret missing `closed_date` values. For example, open or in-progress requests may not have a closed date yet.

## Question 6: What date range does the dataset cover?

This question identifies the earliest and latest request creation dates in the sample. Understanding the date range helps clarify whether the analysis covers a short time window or a longer historical period.

In [59]:
# Date range covered by the dataset
earliest_created_date = df["created_date"].min()
latest_created_date = df["created_date"].max()

print("Earliest created date:", earliest_created_date)
print("Latest created date:", latest_created_date)

Earliest created date: 2026-06-16 19:46:09
Latest created date: 2026-06-26 02:06:32


### Observation

The sample covers NYC 311 requests created between June 16, 2026 and June 26, 2026.

This means the analysis is based on a recent short-term sample, not the full historical NYC 311 dataset.

## Question 7: Which cities appear most often in the dataset?

This question looks at the `city` field to understand which city/location names appear most frequently in the sample. This can help identify geographic concentration and also reveal possible location data quality issues.

In [60]:
# Top 10 cities by number of requests
top_cities = df["city"].value_counts().head(10)

top_cities

city
BROOKLYN         30117
BRONX            18539
NEW YORK         16510
STATEN ISLAND     3722
JAMAICA           2414
ASTORIA           2087
FLUSHING          1681
QUEENS            1505
FRESH MEADOWS     1160
RIDGEWOOD         1095
Name: count, dtype: int64

### Observation

The most common city value in the sample is `BROOKLYN`, followed by `BRONX`, `NEW YORK`, and `STATEN ISLAND`.

Some values in the `city` field are borough names, while others are neighborhood or area names such as `JAMAICA`, `ASTORIA`, `FLUSHING`, and `RIDGEWOOD`. This suggests that the `city` column may not represent one consistent geographic level, which could be reviewed later during data quality checks.

## Question 8: What are the most common complaint descriptors?

This question looks at the `descriptor` field, which provides more specific details about the complaint type. Descriptor analysis helps identify the most common subcategories within broader complaint groups.

In [61]:
# Top 10 complaint descriptors
top_descriptors = df["descriptor"].value_counts().head(10)

top_descriptors

descriptor
Loud Music/Party                 14441
Blocked Hydrant                   4536
Posted Parking Sign Violation     4238
Banging/Pounding                  4008
No Access                         3605
Blocked Sidewalk                  2568
Loud Talking                      2286
Trash                             2159
With License Plate                1979
PESTS                             1784
Name: count, dtype: int64

### Observation

The most common descriptor in this sample is `Loud Music/Party`, with 14,441 records. Other frequent descriptors include parking-related issues such as `Blocked Hydrant` and `Posted Parking Sign Violation`, as well as noise-related details such as `Banging/Pounding` and `Loud Talking`.

This provides more detailed context behind the broader complaint types. For example, many noise complaints appear to be related to loud music, parties, banging, or loud talking.

## Question 9: Which borough has the most noise-related complaints?

This question filters the dataset for noise-related complaint types and then counts those complaints by borough. This helps identify where noise complaints are most concentrated in the sample.

In [62]:
# Noise-related complaints by borough
noise_df = df[df["complaint_type"].str.contains("Noise", case=False, na=False)]

noise_by_borough = noise_df["borough"].value_counts()

noise_by_borough

borough
BROOKLYN         7005
BRONX            6149
QUEENS           5696
MANHATTAN        5180
STATEN ISLAND     705
Unspecified         2
Name: count, dtype: int64

### Observation

Brooklyn has the highest number of noise-related complaints in this sample, with 7,005 records. The Bronx, Queens, and Manhattan also have high noise complaint volumes.

This suggests that noise complaints are a major issue across multiple boroughs, but Brooklyn has the highest count in this sample.

## Question 10: What are the top complaint types in Brooklyn?

This question filters the dataset to Brooklyn records and then identifies the most common complaint types within that borough. This helps compare local complaint patterns against the overall citywide sample.

In [63]:
# Top complaint types in Brooklyn
brooklyn_df = df[df["borough"] == "BROOKLYN"]

brooklyn_top_complaints = brooklyn_df["complaint_type"].value_counts().head(10)

brooklyn_top_complaints

complaint_type
Illegal Parking            6761
Noise - Residential        3579
Noise - Street/Sidewalk    1715
Blocked Driveway           1706
UNSANITARY CONDITION       1069
Street Condition            868
Dirty Condition             697
Noise - Commercial          617
Water System                600
Abandoned Vehicle           592
Name: count, dtype: int64

### Observation

In Brooklyn, the most common complaint type is `Illegal Parking`, with 6,761 records. Noise-related complaints are also highly common, including `Noise - Residential`, `Noise - Street/Sidewalk`, and `Noise - Commercial`.

This shows that Brooklyn’s top complaint patterns are similar to the overall citywide sample, where illegal parking and noise complaints are also major categories.

# Day 2 Summary

In this notebook, I completed basic pandas analysis on a 100,000-row sample of the NYC 311 service requests dataset.

## Work Completed

- Loaded the NYC 311 dataset from the local `data` folder
- Selected individual and multiple columns for analysis
- Used `value_counts()` to summarize complaint types, boroughs, agencies and statuses
- Filtered the dataset for Brooklyn complaints, noise-related complaints and Brooklyn noise complaints
- Sorted records by created date
- Converted `created_date` and `closed_date` from object/text columns to datetime columns
- Answered 10 basic business questions using pandas

## Key Findings

- `Illegal Parking` is the most common complaint type in the sample.
- Noise-related complaints are also highly frequent, especially `Noise - Residential` and `Noise - Street/Sidewalk`.
- Brooklyn has the highest number of 311 requests in the sample.
- NYPD handles the largest number of requests, likely because many high-volume complaint types involve parking and noise issues.
- Most requests are marked as `Closed`, but a significant number are still `Open` or `In Progress`.
- The sample covers requests created between June 16, 2026 and June 26, 2026.
- The `city` column contains a mix of borough names and neighborhood/location names, which may need review during later data quality checks.

## Next Step

The next step is Day 3: missing value analysis. This will involve calculating missing counts and percentages, identifying which missing values are expected versus problematic, and creating the first DataGuard data quality report.